File: ios/ArveyGo/ArveyGo/ViewModels/AuthViewModel.swift
import SwiftUI
import Combine

// MARK: - Auth ViewModel
@MainActor
class AuthViewModel: ObservableObject {
    @Published var isLoggedIn = false
    @Published var isLoading = false
    @Published var currentUser: AppUser?
    @Published var errorMessage: String?

    // Login fields
    @Published var loginEmail = ""
    @Published var loginPassword = ""

    // Register fields
    @Published var registerName = ""
    @Published var registerEmail = ""
    @Published var registerPassword = ""
    @Published var registerPasswordConfirm = ""

    // Forgot password
    @Published var forgotEmail = ""
    @Published var resetSent = false

    // Remember Me
    @Published var rememberMe = false

    // WebSocket config received from server (or generated locally)
    @Published var wsConfig: WSConfig?

    // Hardcoded admin credentials (offline fallback only)
    private let offlineAdminEmail = "admin@admin.com"
    private let offlineAdminPassword = "123"

    // Registered users storage (in-memory for offline demo)
    private var registeredUsers: [AppUser] = []
    private var userPasswords: [String: String] = [:]

    // MARK: - Init — auto-login if token exists
    init() {
        APIService.shared.onSessionExpired = { [weak self] in
            Task { @MainActor in
                self?.logout()
            }
        }
        if APIService.shared.hasStoredToken {
            attemptAutoLogin()
        }
    }

    // MARK: - Auto Login
    /// Try to restore session using stored Bearer token
    private func attemptAutoLogin() {
        isLoading = true
        Task {
            do {
                let user = try await APIService.shared.fetchMe()
                self.currentUser = user
                self.isLoggedIn = true
                self.isLoading = false
                self.generateLocalWSConfig(for: user)
                self.connectWebSocket()
                print("[Auth] Auto-login OK: \(user.name)")
            } catch {
                print("[Auth] Auto-login failed: \(error.localizedDescription)")
                APIService.shared.clearToken()
                self.isLoading = false
            }
        }
    }

    // MARK: - Login
    func login() {
        errorMessage = nil
        guard !loginEmail.isEmpty else {
            errorMessage = "E-posta adresi gerekli"
            return
        }
        guard loginEmail.contains("@") else {
            errorMessage = "Geçerli bir e-posta adresi girin"
            return
        }
        guard !loginPassword.isEmpty else {
            errorMessage = "Şifre gerekli"
            return
        }

        isLoading = true
        loginViaAPI()
    }

    // MARK: - Real API Login
    private func loginViaAPI() {
        Task {
            do {
                let response = try await APIService.shared.login(
                    email: loginEmail,
                    password: loginPassword
                )

                self.currentUser = response.user
                self.isLoggedIn = true
                self.isLoading = false

                // Generate WS config with user info
                self.generateLocalWSConfig(for: response.user)

                // Auto-connect WebSocket
                connectWebSocket()

                // Request push permission & register device token with backend
                registerPushToken()

                print("[Auth] Login OK: \(response.user.name)")

            } catch let error as APIError {
                self.isLoading = false

                switch error {
                case .networkError:
                    // Network failure → offer offline fallback
                    print("[Auth] Network error, trying offline login")
                    self.loginOffline()
                default:
                    self.errorMessage = error.errorDescription
                }

            } catch {
                self.isLoading = false
                // Unknown error → try offline
                print("[Auth] API login failed: \(error.localizedDescription)")
                self.loginOffline()
            }
        }
    }

    // MARK: - Offline / Demo Login (fallback when server is unreachable)
    private func loginOffline() {
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.8) { [weak self] in
            guard let self = self else { return }
            self.isLoading = false

            let email = self.loginEmail.lowercased().trimmingCharacters(in: .whitespaces)

            // Check admin credentials
            if email == self.offlineAdminEmail && self.loginPassword == self.offlineAdminPassword {
                self.currentUser = AppUser.dummy
                self.isLoggedIn = true
                self.generateLocalWSConfig(for: AppUser.dummy)
                self.connectWebSocket()
                return
            }

            // Check registered users
            if let password = self.userPasswords[email],
               password == self.loginPassword,
               let user = self.registeredUsers.first(where: { $0.email == email }) {
                self.currentUser = user
                self.isLoggedIn = true
                self.generateLocalWSConfig(for: user)
                self.connectWebSocket()
                return
            }

            self.errorMessage = "Sunucuya bağlanılamadı. Çevrimdışı giriş: admin@admin.com / 123"
        }
    }

    // MARK: - Token Refresh
    func refreshTokenIfNeeded() {
        Task {
            do {
                _ = try await APIService.shared.refreshToken()
                print("[Auth] Token refreshed")
            } catch {
                print("[Auth] Token refresh failed: \(error.localizedDescription)")
                // If refresh fails with 401, force re-login
                if case APIError.unauthorized = error {
                    logout()
                }
            }
        }
    }

    // MARK: - WebSocket Configuration
    private func generateLocalWSConfig(for user: AppUser) {
        let jwt = JWTHelper.issueLiveMapToken(
            sub: user.id,
            companyId: user.companyId
        )
        wsConfig = WSConfig(
            url: AppConfig.wsURL,
            token: jwt,
            pingInterval: Int(AppConfig.wsPingInterval)
        )
    }

    func connectWebSocket() {
        guard let config = wsConfig else { return }
        WebSocketManager.shared.connect(url: config.url, token: config.token)
    }

    func disconnectWebSocket() {
        WebSocketManager.shared.disconnect()
    }

    // MARK: - Push Token Registration
    /// Request push permission, then send device token to backend
    private func registerPushToken() {
        AppDelegate.requestPushPermission()

        // Observe token changes — when Apple delivers the token, send it to backend
        var cancellable: AnyCancellable?
        cancellable = DeviceTokenStore.shared.$token
            .compactMap { $0 }          // wait for non-nil
            .first()                     // only once
            .sink { token in
                Task {
                    await APIService.shared.registerPushToken(token)
                }
                cancellable?.cancel()
            }

        // If token is already available (e.g. from previous session), send immediately
        if let existingToken = DeviceTokenStore.shared.token {
            Task {
                await APIService.shared.registerPushToken(existingToken)
            }
            cancellable?.cancel()
        }
    }

    // MARK: - Login via Phone + OTP
    func loginWithOTP(phone: String, otp: String, completion: @escaping (Bool) -> Void) {
        errorMessage = nil
        isLoading = true

        DispatchQueue.main.asyncAfter(deadline: .now() + 0.8) { [weak self] in
            guard let self = self else { return }
            self.isLoading = false

            if otp == "000000" {
                self.currentUser = AppUser.dummy
                self.isLoggedIn = true
                self.generateLocalWSConfig(for: AppUser.dummy)
                self.connectWebSocket()
                completion(true)
            } else {
                completion(false)
            }
        }
    }

    // MARK: - Register
    func register() {
        errorMessage = nil

        guard !registerName.isEmpty else { errorMessage = "Ad Soyad gerekli"; return }
        guard !registerEmail.isEmpty, registerEmail.contains("@") else { errorMessage = "Geçerli bir e-posta adresi girin"; return }
        guard registerPassword.count >= 8 else { errorMessage = "Şifre en az 8 karakter olmalı"; return }
        guard registerPassword == registerPasswordConfirm else { errorMessage = "Şifreler eşleşmiyor"; return }

        isLoading = true

        DispatchQueue.main.asyncAfter(deadline: .now() + 1.2) { [weak self] in
            guard let self = self else { return }
            self.isLoading = false

            let email = self.registerEmail.lowercased().trimmingCharacters(in: .whitespaces)
            let name = self.registerName.trimmingCharacters(in: .whitespaces)

            let initials = name.split(separator: " ")
                .prefix(2)
                .compactMap { $0.first.map { String($0).uppercased() } }
                .joined()

            let newUser = AppUser(
                id: UUID().uuidString,
                name: name,
                email: email,
                avatar: initials.isEmpty ? "U" : initials,
                role: "Kullanıcı",
                roleKey: "user",
                companyId: 1
            )

            self.registeredUsers.append(newUser)
            self.userPasswords[email] = self.registerPassword
            self.currentUser = newUser
            self.isLoggedIn = true
        }
    }

    // MARK: - Forgot Password
    func sendResetLink() {
        errorMessage = nil
        guard !forgotEmail.isEmpty, forgotEmail.contains("@") else {
            errorMessage = "Geçerli bir e-posta adresi girin"
            return
        }

        isLoading = true
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            guard let self = self else { return }
            self.isLoading = false
            self.resetSent = true
        }
    }

    // MARK: - Logout
    func logout() {
        disconnectWebSocket()
        wsConfig = nil

        // Logout from API (clears token)
        Task { await APIService.shared.logout() }

        withAnimation {
            isLoggedIn = false
            currentUser = nil
            loginEmail = ""
            loginPassword = ""
            errorMessage = nil
        }
    }

    // MARK: - Clear
    func clearLoginFields() {
        loginEmail = ""
        loginPassword = ""
        errorMessage = nil
    }

    func clearRegisterFields() {
        registerName = ""
        registerEmail = ""
        registerPassword = ""
        registerPasswordConfirm = ""
        errorMessage = nil
    }

    func clearForgotFields() {
        forgotEmail = ""
        errorMessage = nil
        resetSent = false
    }
}

File: android/app/src/main/java/com/arveya/arveygo/viewmodels/AuthViewModel.kt
package com.arveya.arveygo.viewmodels

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.arveya.arveygo.models.AppUser
import com.arveya.arveygo.models.WSConfig
import com.arveya.arveygo.services.*
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class AuthViewModel(application: Application) : AndroidViewModel(application) {

    private val ctx get() = getApplication<Application>().applicationContext

    private val _isLoggedIn = MutableStateFlow(false)
    val isLoggedIn: StateFlow<Boolean> = _isLoggedIn

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _currentUser = MutableStateFlow<AppUser?>(null)
    val currentUser: StateFlow<AppUser?> = _currentUser

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    // Login fields
    val loginEmail = MutableStateFlow("")
    val loginPassword = MutableStateFlow("")

    // Register fields
    val registerName = MutableStateFlow("")
    val registerEmail = MutableStateFlow("")
    val registerPassword = MutableStateFlow("")
    val registerPasswordConfirm = MutableStateFlow("")

    // Forgot password
    val forgotEmail = MutableStateFlow("")
    private val _resetSent = MutableStateFlow(false)
    val resetSent: StateFlow<Boolean> = _resetSent

    // Remember Me
    val rememberMe = MutableStateFlow(false)

    // WS config
    private var wsConfig: WSConfig? = null

    // Offline fallback credentials
    private val offlineAdminEmail = "admin@admin.com"
    private val offlineAdminPassword = "123"

    // Registered users (in-memory for offline demo)
    private val registeredUsers = mutableListOf<AppUser>()
    private val userPasswords = mutableMapOf<String, String>()

    // MARK: - Init — restore token & auto-login
    init {
        APIService.initialize(ctx)
        // Register global 401 handler — any API call that returns 401 will auto-logout
        APIService.onSessionExpired = { logout() }
        if (APIService.hasStoredToken) {
            attemptAutoLogin()
        }
    }

    private fun attemptAutoLogin() {
        _isLoading.value = true
        viewModelScope.launch {
            try {
                val user = APIService.fetchMe()
                _currentUser.value = user
                _isLoggedIn.value = true
                _isLoading.value = false
                generateLocalWSConfig(user)
                connectWebSocket()
                Log.d("Auth", "Auto-login OK: ${user.name}")
            } catch (e: Exception) {
                Log.d("Auth", "Auto-login failed: ${e.localizedMessage}")
                APIService.clearToken(ctx)
                _isLoading.value = false
            }
        }
    }

    // MARK: - Login
    fun login() {
        _errorMessage.value = null
        val email = loginEmail.value.trim()
        val password = loginPassword.value

        if (email.isEmpty()) { _errorMessage.value = "E-posta adresi gerekli"; return }
        if (!email.contains("@")) { _errorMessage.value = "Geçerli bir e-posta adresi girin"; return }
        if (password.isEmpty()) { _errorMessage.value = "Şifre gerekli"; return }

        _isLoading.value = true
        loginViaAPI(email, password)
    }

    private fun loginViaAPI(email: String, password: String) {
        viewModelScope.launch {
            try {
                val response = APIService.login(email, password, ctx)
                _currentUser.value = response.user
                _isLoggedIn.value = true
                _isLoading.value = false

                generateLocalWSConfig(response.user)
                connectWebSocket()

                Log.d("Auth", "Login OK: ${response.user.name}")

            } catch (e: APIException.NetworkError) {
                _isLoading.value = false
                Log.d("Auth", "Network error, trying offline: ${e.localizedMessage}")
                loginOffline(email, password)
            } catch (e: APIException) {
                _isLoading.value = false
                _errorMessage.value = e.message
            } catch (e: Exception) {
                _isLoading.value = false
                Log.d("Auth", "API login failed: ${e.localizedMessage}")
                loginOffline(email, password)
            }
        }
    }

    // MARK: - Offline / Demo Login (fallback when server is unreachable)
    private fun loginOffline(email: String, password: String) {
        viewModelScope.launch {
            delay(800)
            _isLoading.value = false
            val em = email.lowercase().trim()

            if (em == offlineAdminEmail && password == offlineAdminPassword) {
                _currentUser.value = AppUser.dummy
                _isLoggedIn.value = true
                generateLocalWSConfig(AppUser.dummy)
                connectWebSocket()
                return@launch
            }

            val storedPassword = userPasswords[em]
            if (storedPassword == password) {
                val user = registeredUsers.firstOrNull { it.email == em }
                if (user != null) {
                    _currentUser.value = user
                    _isLoggedIn.value = true
                    generateLocalWSConfig(user)
                    connectWebSocket()
                    return@launch
                }
            }

            _errorMessage.value = "Sunucuya bağlanılamadı. Çevrimdışı giriş: admin@admin.com / 123"
        }
    }

    // MARK: - Token Refresh
    fun refreshTokenIfNeeded() {
        viewModelScope.launch {
            try {
                APIService.refreshToken(ctx)
                Log.d("Auth", "Token refreshed")
            } catch (e: Exception) {
                Log.d("Auth", "Token refresh failed: ${e.localizedMessage}")
                if (e is APIException.Unauthorized) {
                    logout()
                }
            }
        }
    }

    // MARK: - WS Configuration
    private fun generateLocalWSConfig(user: AppUser) {
        val jwt = JWTHelper.issueLiveMapToken(sub = user.id, companyId = user.companyId)
        wsConfig = WSConfig(url = AppConfig.WS_URL, token = jwt, pingInterval = AppConfig.WS_PING_INTERVAL.toInt())
        Log.d("WS", "generateLocalWSConfig: url=${AppConfig.WS_URL}, userId=${user.id}")
    }

    fun connectWebSocket() {
        wsConfig?.let {
            WebSocketManager.connect(it.url, it.token)
        }
    }

    fun disconnectWebSocket() {
        WebSocketManager.disconnect()
    }

    // MARK: - Login via Phone + OTP
    fun loginWithOTP(phone: String, otp: String, onResult: (Boolean) -> Unit) {
        _errorMessage.value = null
        _isLoading.value = true

        viewModelScope.launch {
            delay(800)
            _isLoading.value = false

            if (otp == "000000") {
                _currentUser.value = AppUser.dummy
                _isLoggedIn.value = true
                generateLocalWSConfig(AppUser.dummy)
                connectWebSocket()
                onResult(true)
            } else {
                onResult(false)
            }
        }
    }

    // MARK: - Register
    fun register() {
        _errorMessage.value = null
        val name = registerName.value.trim()
        val email = registerEmail.value.trim().lowercase()
        val password = registerPassword.value
        val confirm = registerPasswordConfirm.value

        if (name.isEmpty()) { _errorMessage.value = "Ad Soyad gerekli"; return }
        if (email.isEmpty() || !email.contains("@")) { _errorMessage.value = "Geçerli bir e-posta adresi girin"; return }
        if (password.length < 8) { _errorMessage.value = "Şifre en az 8 karakter olmalı"; return }
        if (password != confirm) { _errorMessage.value = "Şifreler eşleşmiyor"; return }

        _isLoading.value = true

        viewModelScope.launch {
            delay(1200)
            _isLoading.value = false

            val initials = name.split(" ").take(2).mapNotNull { it.firstOrNull()?.uppercase() }.joinToString("")
            val newUser = AppUser(
                id = java.util.UUID.randomUUID().toString(),
                name = name, email = email,
                avatar = initials.ifEmpty { "U" },
                role = "Kullanıcı", roleKey = "user", companyId = 1
            )
            registeredUsers.add(newUser)
            userPasswords[email] = password
            _currentUser.value = newUser
            _isLoggedIn.value = true
        }
    }

    // MARK: - Forgot Password
    fun sendResetLink() {
        _errorMessage.value = null
        val email = forgotEmail.value.trim()
        if (email.isEmpty() || !email.contains("@")) { _errorMessage.value = "Geçerli bir e-posta adresi girin"; return }

        _isLoading.value = true
        viewModelScope.launch {
            delay(1000)
            _isLoading.value = false
            _resetSent.value = true
        }
    }

    // MARK: - Logout
    fun logout() {
        disconnectWebSocket()
        wsConfig = null

        viewModelScope.launch { APIService.logout(ctx) }

        _isLoggedIn.value = false
        _currentUser.value = null
        loginEmail.value = ""
        loginPassword.value = ""
        _errorMessage.value = null
    }

    // MARK: - Clear
    fun clearLoginFields() {
        loginEmail.value = ""
        loginPassword.value = ""
        _errorMessage.value = null
    }

    fun clearRegisterFields() {
        registerName.value = ""
        registerEmail.value = ""
        registerPassword.value = ""
        registerPasswordConfirm.value = ""
        _errorMessage.value = null
    }

    fun clearForgotFields() {
        forgotEmail.value = ""
        _errorMessage.value = null
        _resetSent.value = false
    }
}
